src/Stratis.Bitcoin/Base/BackgroundWork/BackgroundWorkProvider.cs

```diff
@@ -0,0 +1,214 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Base.BackgroundWork
+{
+    /// <summary>
+    /// Provides functionality for creating and tracking operations that happens in background.
+    /// </summary>
+    public partial class BackgroundWorkProvider : IBackgroundWorkProvider
+    {
+        private const int DefaultLoopRepeatInterval = 1000;
+
+        private object lockAsyncDelegates;
```

Comment:
This are rearly called we can reuse the lock on both dictionaries

Reply:
this costs just an object instantiation, isn't better?

Reply:
ok yeah